### PR TITLE
Pass room metadata from 3pid invite emails all the way through to the…

### PR DESCRIPTION
**Pull request by dbkr**
… relevant components so we can display it.

### CODE REVIEW

**Comment by dbkr** [comment]
Fixes https://github.com/vector-im/vector-web/issues/1006

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -361,6 +365,7 @@ module.exports = React.createClass({
                         room_id: result.room_id,
                         event_id: payload.event_id,
                         invite_sign_url: payload.invite_sign_url,
+                        room_oob_data: payload.room_oob_data,
```
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: This looks inconsistent. s/room_//g ?

This highlights a general problem here - these payloads, and the arguments to `_viewRoom` are getting a bit silly. I feel like they should be factored out to a separate object which documents its fields.

File: src/components/structures/MatrixChat.js [hunk 2]
```diff
@@ -448,7 +453,10 @@ module.exports = React.createClass({
     //
     // eventId is optional and will cause a switch to the context of that
     // particular event.
-    _viewRoom: function(roomId, showSettings, eventId, invite_sign_url) {
+    // @param {Object} room_oob_data Object of additional data about the room
```
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: s/room_//

File: src/components/structures/MatrixChat.js [hunk 3]
```diff
@@ -552,10 +561,12 @@ module.exports = React.createClass({
                     room_alias: self.starting_room_alias,
                     event_id: self.starting_event_id,
                     invite_sign_url: self.starting_room_invite_sign_url,
+                    room_oob_data: self.starting_room_oob_data,
```
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: s/room_// on the key name ?

File: src/components/structures/MatrixChat.js [hunk 4]
```diff
@@ -682,11 +700,16 @@ module.exports = React.createClass({
                         room_alias: roomString,
                         event_id: eventId,
                         invite_sign_url: params.signurl,
+                        oob_data: oob_data,
                     });
                 } else {
                     // Okay, we'll take you here soon...
+                    // XXX: There are way too many of these:
+                    // It would probably be better to handle whether the SDK is
+                    // ready or not in the view_room_alias handler instead.
```
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: or at least stick them in a different object

File: src/components/structures/RoomView.js [hunk 5]
```diff
@@ -59,6 +59,12 @@ module.exports = React.createClass({
         // (given as part of the link in the invite email)
         inviteSignUrl: React.PropTypes.string,
 
+        // Any data about the room that would normally come from the Home Server
+        // but has been passed out-of-band, eg. the room name and avatar URL
+        // from an email invite (a workaround for the fact that we can't
+        // get this information from the HS using an email invite).
+        oobData: React.PropTypes.object,
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: I'd like to see the fields of this object documented.

File: src/components/structures/RoomView.js [hunk 6]
```diff
@@ -1293,6 +1305,7 @@ module.exports = React.createClass({
         return (
             <div className={ "mx_RoomView" + (inCall ? " mx_RoomView_inCall" : "") } ref="roomView">
                 <RoomHeader ref="header" room={this.state.room} searchInfo={searchInfo}
+                    oobData={this.state.oobData}
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: s/state/props/ ?

File: src/components/views/rooms/RoomHeader.js [hunk 7]
```diff
@@ -217,18 +218,27 @@ module.exports = React.createClass({
                 }
 
                 // XXX: this is a bit inefficient - we could just compare room.name for 'Empty room'...
-                var members = this.props.room.getJoinedMembers();
```
Comment (richvdh) [inline src/components/views/rooms/RoomHeader.js]: long method is long :( :(

File: src/components/views/rooms/RoomHeader.js [hunk 8]
```diff
@@ -217,18 +218,27 @@ module.exports = React.createClass({
...
+                            settingsHint = true;
+                        }
                     }
                 }
 
+                var roomName = 'Join Room';
+                if (this.props.oobData && this.props.oobData.name) {
+                    roomName = this.props.oobData.name;
+                } else if (this.props.room) {
+                    this.props.room.name;
```
Comment (richvdh) [inline src/components/views/rooms/RoomHeader.js]: missing `roomName =`

File: src/components/views/rooms/RoomHeader.js [hunk 9]
```diff
@@ -217,18 +218,27 @@ module.exports = React.createClass({
                 }
 
                 // XXX: this is a bit inefficient - we could just compare room.name for 'Empty room'...
-                var members = this.props.room.getJoinedMembers();
                 var settingsHint = false;
-                if (members.length === 1 && members[0].userId === MatrixClientPeg.get().credentials.userId) {
-                    var name = this.props.room.currentState.getStateEvents('m.room.name', '');
-                    if (!name || !name.getContent().name) {
-                        settingsHint = true;
+                var members = this.props.room ? this.props.room.getJoinedMembers() : undefined;
```
Comment (richvdh) [inline src/components/views/rooms/RoomHeader.js]: it looks like it was probably wrong before, but:

why can we no longer rely on `this.props.room` being set, when we could before?

File: src/components/views/avatars/RoomAvatar.js [hunk 10]
```diff
@@ -22,7 +23,8 @@ module.exports = React.createClass({
     displayName: 'RoomAvatar',
 
     propTypes: {
-        room: React.PropTypes.object.isRequired,
+        room: React.PropTypes.object,
```
Comment (richvdh) [inline src/components/views/avatars/RoomAvatar.js]: it would be nice to add some comments to explain how `room` and `oobData` interact. What if I set both? What if I set neither?

**Comment by richvdh** [comment]
This code is hard to follow, and you're not making it any better :/. I won't insist on a refactorathon, but I'd encourage you to do what you can to consider the guy who comes after you. He probably knows where you live.

File: src/components/views/rooms/RoomHeader.js (hunk 9 above)
Comment (dbkr) [inline src/components/views/rooms/RoomHeader.js]: Previously the only case where room wasn't set was if simpleHeader was passed, in which case we never got here.

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: The reason I used the room_prefix here is because to clarify that it pertains to the room, since it's in MatrixChat, where in RoomView and below, they only handle the room so it's obvious. I don't mind though.

File: src/components/views/rooms/RoomHeader.js (hunk 7 above)
Comment (dbkr) [inline src/components/views/rooms/RoomHeader.js]: Yes, this could be a great case for a separate logic class.

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: ok, but things read the `view_room` payload and look for `oob_data`, not `room_oob_data`. This one looks like a bug rather than just being inconsistent.

**Comment by dbkr** [comment]
OK, this should be a little better. Happy to ditch the room_ if you think that's clearer. ptal.

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: still broken?

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: still wrong, though it's only comments so doesn't *technically* matter

File: src/components/structures/MatrixChat.js (hunk 3 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: this is read as `oob_data` at line :356, so needs fixing

File: src/components/structures/RoomView.js [hunk 11]
```diff
@@ -59,6 +59,15 @@ module.exports = React.createClass({
         // (given as part of the link in the invite email)
         inviteSignUrl: React.PropTypes.string,
 
+        // Any data about the room that would normally come from the Home Server
+        // but has been passed out-of-band, eg. the room name and avatar URL
+        // from an email invite (a workaround for the fact that we can't
+        // get this information from the HS using an email invite).
+        // Fields:
+        //  * name (string) The room's name
+        //  * avatarUrl (string) The mxc:// avatar URL for the room
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: `inviterName` ?

File: src/components/structures/RoomView.js (hunk 5 above)
Comment (richvdh) [inline src/components/structures/RoomView.js]: done

File: src/components/structures/RoomView.js (hunk 6 above)
Comment (richvdh) [inline src/components/structures/RoomView.js]: done

File: src/components/views/rooms/RoomHeader.js (hunk 8 above)
Comment (richvdh) [inline src/components/views/rooms/RoomHeader.js]: done

File: src/components/views/rooms/RoomHeader.js (hunk 9 above)
Comment (richvdh) [inline src/components/views/rooms/RoomHeader.js]: ok.

done

File: src/components/structures/RoomView.js [hunk 12]
```diff
@@ -1052,13 +1061,19 @@ module.exports = React.createClass({
...
                 }
                 else {
+                    var inviterName = undefined;
+                    if (this.props.oobData) {
+                        inviterName = this.props.oobData.inviterName;
+                    }
+
                     return (
                         <div className="mx_RoomView">
-                            <RoomHeader ref="header" room={this.state.room} simpleHeader="Join room"/>
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: What about the case where there is no oobData? do we not still need simpleHeader?

File: src/components/views/rooms/RoomHeader.js (hunk 7 above)
Comment (richvdh) [inline src/components/views/rooms/RoomHeader.js]: or even just some separate methods

File: src/components/views/avatars/RoomAvatar.js (hunk 10 above)
Comment (richvdh) [inline src/components/views/avatars/RoomAvatar.js]: done

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: Oops, you're absolutely right, thank you.

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: Ah yes, I se what you mean now. Thanks.

File: src/components/structures/MatrixChat.js (hunk 3 above)
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: Right you are.

File: src/components/structures/RoomView.js (hunk 12 above)
Comment (dbkr) [inline src/components/structures/RoomView.js]: I made RoomHeader drop back to displaying the simple, 'Join Room' message if it really can't get any information about the room, which should make it redundant here (I'd be in favour of splitting out the simple header stuff into a separate component that's used by RoomHeader and then directly in the other places that use RoomHeader's simpleHeader).

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: *still* broken.

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: done

File: src/components/structures/MatrixChat.js (hunk 3 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: done

File: src/components/structures/RoomView.js (hunk 11 above)
Comment (richvdh) [inline src/components/structures/RoomView.js]: done

File: src/components/structures/RoomView.js (hunk 12 above)
Comment (richvdh) [inline src/components/structures/RoomView.js]: Right, yes. Ugh. So would I. 

Is simpleHeader used anywhere now? Or should it be killed?

**Comment by richvdh** [comment]
I think you still have an excess `room_` at MatrixChat:368. Other than that, lgtm for an initial merge to develop

**Comment by dbkr** [comment]
ah yes, so I do. Fixed.

File: src/components/structures/RoomView.js (hunk 12 above)
Comment (dbkr) [inline src/components/structures/RoomView.js]: It's referenced by the Create Room screen which is no longer hooked up to anything, and the user settings screen (which annoys me because settings is not a room).
